pkg/nodeset/v1alpha1/types.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+type NodeClassResourceType string
+
+const (
+	NodeClassResourceFile      NodeClassResourceType = "File"
+	NodeClassResourceReference NodeClassResourceType = "Reference"
+)
+
+type NodeClassResource struct {
+	// Type is the type of the resource
+	Type NodeClassResourceType `json:"type,omitempty"`
+
+	// Name is the name of the resource
+	Name string `json:"name,omitempty"`
+
+	// For File type
+	// Path is the path of the file
+	// +optional
+	Path string `json:"path,omitempty"`
+	// Owner is the owner of the file
+	// +optional
+	Owner string `json:"owner,omityempty"`
+	// Permission is the permission of the file
+	// +optional
+	Permission string `json:"permission,omityempty"`
+	// Template is the template or content of the file
+	// +optional
+	Template string `json:"template,omitempty"`
+
+	// For Reference type
+	// Reference points to an Object in registry that the controller will use when
+	// provisioning the node
+	// +optional
+	Reference *v1.ObjectReference `json:"reference,omitempty"`
+}
+
+// NodeClass describes the parameters for a class of nodes that can be
+// provisioned by the specific controller
+type NodeClass struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// NodeController specifes which node controller should provision and
+	// control this class of nodes
+	NodeController string `json:"nodeController,omitempty"`
```

Comment:
Yes. And I'm thinking is it really necessary to have separate NodeController and NodeSetController fields instead of a Backend field? I'm under the impression that these two controllers should always work together seamlessly.  And for that they will be mostly likely to come from the same backend. How high is the chance of two controllers from different backends to work together?